notebook/terminal/terminalmanager.py

```diff
@@ -0,0 +1,150 @@
+    # -------------------------------------------------------------------------
+    def __init__(self, *args, **kwargs):
+        super(TerminalManager, self).__init__(*args, **kwargs)
+
+    def create(self):
+        """Create a new terminal."""
+        name, term = self.new_named_terminal()
+        # Monkey-patch last-activity, similar to kernels.  Should we need
+        # more functionality per terminal, we can look into possible sub-
+        # classing or containment then.
+        term.last_activity = utcnow()
+        model = self.terminal_model(name)
+        # Increase the metric by one because a new terminal was created
+        TERMINAL_CURRENTLY_RUNNING_TOTAL.inc()
+        # Ensure culler is initialized
+        self._initialize_culler()
+        return model
+
+    def get(self, name):
+        """Get terminal 'name'."""
+        model = self.terminal_model(name)
+        return model
+
+    def list(self):
+        """Get a list of all running terminals."""
+        models = [self.terminal_model(name) for name in self.terminals]
+
+        # Update the metric below to the length of the list 'terms'
+        TERMINAL_CURRENTLY_RUNNING_TOTAL.set(
+            len(models)
+        )
+        return models
+
+    async def terminate(self, name, force=False):
+        """Terminate terminal 'name'."""
+        self._check_terminal(name)
+        await super(TerminalManager, self).terminate(name, force=force)
+
+        # Decrease the metric below by one
+        # because a terminal has been shutdown
+        TERMINAL_CURRENTLY_RUNNING_TOTAL.dec()
+
+    async def terminate_all(self):
+        """Terminate all terminals."""
+        terms = [name for name in self.terminals]
+        for term in terms:
+            await self.terminate(term, force=True)
+
+    def terminal_model(self, name):
+        """Return a JSON-safe dict representing a terminal.
+        For use in representing terminals in the JSON APIs.
+        """
+        self._check_terminal(name)
+        term = self.terminals[name]
+        model = {
+            "name": name,
+            "last_activity": isoformat(term.last_activity),
+        }
+        return model
+
+    def _check_terminal(self, name):
+        """Check a that terminal 'name' exists and raise 404 if not."""
+        if name not in self.terminals:
+            raise web.HTTPError(404, u'Terminal not found: %s' % name)
+
+    def _initialize_culler(self):
+        """Start culler if 'cull_inactive_timeout' is greater than zero.
+        Regardless of that value, set flag that we've been here.
+        """
+        if not self._initialized_culler and self.cull_inactive_timeout > 0:
+            if self._culler_callback is None:
+                loop = IOLoop.current()
+                if self.cull_interval <= 0:  # handle case where user set invalid value
+                    self.log.warning("Invalid value for 'cull_interval' detected (%s) - using default value (%s).",
```

Comment:
Hm, this looks like a type issue that we should catch at the trait definition, no? Should we catch this with a `@validate` call at the top of the class, rather than wait for a terminal to be initialized.

Reply:
Good idea (same should probably happen with the kernel culling stuff as well - from which this was derived). I wasn't able to get logging to work from the validation presumably because it's not initialized(??), so I used warnings.warn.  What's the correct way to do that from validate methods?